pipeline/data_store.py

```diff
@@ -277,6 +211,122 @@ def parse_args(self, *args, **kwargs):
+        self.wcs = None  # astrometric solution
+        self.zp = None  # photometric calibration
+        self.reference = None  # the Reference object needed to make subtractions
+        self.sub_image = None  # subtracted image
+        self.detections = None  # a SourceList object for sources detected in the subtraction image
+        self.cutouts = None  # cutouts around sources
+        self.measurements = None  # photometry and other measurements for each source
+        self.objects = None  # a list of Object associations of Measurements
+
+        # these need to be added to the products_to_clear list
+        self.ref_image = None  # to be used to make subtractions
+        self.sub_image = None  # subtracted image
+        self.reference = None  # the Reference object needed to make subtractions
+        self.exposure_id = None  # use this and section_id to find the raw image
+        self.section_id = None  # corresponds to SensorSection.identifier (*not* .id)
+        self.image_id = None  # use this to specify an image already in the database
+
+        self.warnings_list = None  # will be replaced by a list of warning objects in top_level.Pipeline.run()
+        self.exception = None  # the exception object (so we can re-raise it if needed)
+        self.runtimes = {}  # for each process step, the total runtime in seconds
+        self.memory_usages = {}  # for each process step, the peak memory usage in MB
+        self.products_committed = ''  # a comma separated list of object names (e.g., "image, sources") saved to DB
+        self.report = None  # keep a reference to the report object for this run
+
+        # The database session parsed in parse_args; it could still be None even after parse_args
+        self.session = None
+        self.parse_args(*args, **kwargs)
+
+    @property
+    def exposure( self ):
+        if self._exposure is None:
+            if self.exposure_id is not None:
+                self._exposure = self.get_raw_exposure( session=self.session )
+        return self._exposure
+
+    @exposure.setter
+    def exposure( self, value ):
+        self._exposure = value
+        self.exposure_id = value.id if value is not None else None
+
+    @property
+    def section( self ):
+        if self._section is None:
+            if self.section_id is not None:
+                if self.exposure is not None:
+                    self.exposure.instrument_object.fetch_sections()
+                    self._section = self.exposure.instrument_object.get_section( self.section_id )
+        return self._section
+
+    @property
+    def ref_image( self ):
+        if self.reference is not None:
+            return self.reference.image
+        return None
+
+    @ref_image.setter
+    def ref_image( self, value ):
+        if self.reference is None:
+            self.reference = Reference()
+        self.reference.image = value
+
+    def __getattr__(self, key):
```

Comment:
Python question : is this function guaranteed to be called before (say) the ref_image @property function is called when you do `ds.ref_image`?

...looking at docs, it seems that the case is No, that this only gets called if there's an AttributeError.

Did you want `__getattribute__`?

(Based on the comment, I'm assuming you want this to be called on *any* attribute access.)

Reply:
Test code:
```
class Thing:
    @property
    def twox( self ):
        return 2. * self.val
    
    def __init__( self, val ):
        self.val = val

    def __getattr__( self, attr ):
        print( "__getattr__ called" )

thing = Thing( 5 )
print( thing.val )
print( thing.twox )
print( thing.puppy )
```

prints:

```
5
10.0
__getattr__ called
None
```

but if I understand your comment correctly, you would have wanted "`__getattr__ called`" printed before the lines with 5 and 10.0.

Reply:
```
class Thing:
    @property
    def twox( self ):
        return 2. * self.val
    
    def __init__( self, val ):
        self.val = val

    def __getattr__( self, attr ):
        print( "__getattr__ called" )

    def __getattribute__( self, attr ):
        print( "__getattribute__ called" )
        return super().__getattribute__( attr )
        
thing = Thing( 5 )
print( thing.val )
print( thing.twox )
print( thing.puppy )
```

writes

```
__getattribute__ called
5
__getattribute__ called
__getattribute__ called
10.0
__getattribute__ called
__getattr__ called
None
```

So I think `__getattribute__` is what you want.  (I'm a little bit sad of the two calls before 10.0, but it makes sense why it happens: one for the `.twox`, and one for the `self.val` inside `twox()`.

Reply:
You are correct!